Add check for Go version

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Description

JSII uses the `embed` Go package which was introduced in Go 1.16.

If the Go version is below that, the following error appears:
```
$ cdktf init --template="go" --local
Note: By supplying '--local' option you have chosen local storage mode for storing the state of your stack.
This means that your Terraform state file will be stored locally on disk in a file 'terraform.tfstate' in the root of your project.

We will now set up the project. Please enter the details for your project.
If you want to exit, press ^C.

Project Name: (default: 'learn-cdktf-go') 
Project Description: (default: 'A simple getting started project for cdktf.') 
go: downloading github.com/hashicorp/terraform-cdk-go/cdktf v0.4.0
go: downloading github.com/aws/constructs-go/constructs/v3 v3.3.75
go: downloading github.com/aws/jsii-runtime-go v1.29.0
go: downloading github.com/Masterminds/semver/v3 v3.1.1
../../../go/pkg/mod/github.com/aws/constructs-go/constructs/v3@v3.3.75/jsii/jsii.go:8:2: package embed is not in GOROOT (/usr/local/Cellar/go/1.15.5/libexec/src/embed)
Error: Command failed: go get
    at checkExecSyncError (child_process.js:621:11)
    at execSync (child_process.js:657:15)
    at Object.exports.post (/usr/local/lib/node_modules/cdktf-cli/templates/go/.hooks.sscaff.js:47:3)
    at executePostHook (/usr/local/lib/node_modules/cdktf-cli/node_modules/sscaff/lib/sscaff.js:61:37)
    at Object.sscaff (/usr/local/lib/node_modules/cdktf-cli/node_modules/sscaff/lib/sscaff.js:28:15)
    at async Object.handler (/usr/local/lib/node_modules/cdktf-cli/bin/cmds/init.js:104:9)

Debug Information:
    Terraform CDK version: 0.4.0
    Terraform version: 0.15.0

```
<!--- Please leave a helpful description of the feature request here. --->

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->

We should check the Go version to be able to inform users that they need at least Go 1.16. (It is mentioned in the docs, but one can quickly miss that.)

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Vendor blog posts or documentation?
--->

--- a/packages/cdktf-cli/bin/cmds/deploy.ts
+++ b/packages/cdktf-cli/bin/cmds/deploy.ts
@@ -5,6 +5,7 @@ import { readConfigSync } from "../../lib/config";
 import { renderInk } from "./helper/render-ink";
 import { displayVersionMessage } from "./helper/version-check";
 import { throwIfNotProjectDirectory } from "./helper/check-directory";
+import { checkEnvironment } from "./helper/check-environment";
 
 const config = readConfigSync();
 
@@ -42,6 +43,7 @@ class Command implements yargs.CommandModule {
   public async handler(argv: any) {
     throwIfNotProjectDirectory("deploy");
     await displayVersionMessage();
+    await checkEnvironment("deploy");
     const command = argv.app;
     const outdir = argv.output;
     const autoApprove = argv.autoApprove;

--- a/packages/cdktf-cli/bin/cmds/destroy.ts
+++ b/packages/cdktf-cli/bin/cmds/destroy.ts
@@ -5,6 +5,7 @@ import { readConfigSync } from "../../lib/config";
 import { renderInk } from "./helper/render-ink";
 import { displayVersionMessage } from "./helper/version-check";
 import { throwIfNotProjectDirectory } from "./helper/check-directory";
+import { checkEnvironment } from "./helper/check-environment";
 
 const config = readConfigSync();
 
@@ -41,6 +42,7 @@ class Command implements yargs.CommandModule {
   public async handler(argv: any) {
     throwIfNotProjectDirectory("destroy");
     await displayVersionMessage();
+    await checkEnvironment("destroy");
     const command = argv.app;
     const outdir = argv.output;
     const autoApprove = argv.autoApprove;

--- a/packages/cdktf-cli/bin/cmds/diff.ts
+++ b/packages/cdktf-cli/bin/cmds/diff.ts
@@ -5,6 +5,7 @@ import { readConfigSync } from "../../lib/config";
 import { renderInk } from "./helper/render-ink";
 import { displayVersionMessage } from "./helper/version-check";
 import { throwIfNotProjectDirectory } from "./helper/check-directory";
+import { checkEnvironment } from "./helper/check-environment";
 
 const config = readConfigSync();
 
@@ -37,6 +38,7 @@ class Command implements yargs.CommandModule {
   public async handler(argv: any) {
     throwIfNotProjectDirectory("diff");
     await displayVersionMessage();
+    await checkEnvironment("diff");
     const command = argv.app;
     const outdir = argv.output;
     const stack = argv.stack;

--- a/packages/cdktf-cli/bin/cmds/get.ts
+++ b/packages/cdktf-cli/bin/cmds/get.ts
@@ -9,6 +9,7 @@ import { Get } from "./ui/get";
 import { renderInk } from "./helper/render-ink";
 import { displayVersionMessage } from "./helper/version-check";
 import { throwIfNotProjectDirectory } from "./helper/check-directory";
+import { checkEnvironment } from "./helper/check-environment";
 
 const config = readConfigSync();
 
@@ -43,6 +44,7 @@ class Command implements yargs.CommandModule {
   public async handler(argv: any) {
     throwIfNotProjectDirectory("get");
     await displayVersionMessage();
+    await checkEnvironment("get");
     const args = argv as Arguments;
     const providers = config.terraformProviders ?? [];
     const modules = config.terraformModules ?? [];

--- a/packages/cdktf-cli/bin/cmds/helper/check-environment.ts
+++ b/packages/cdktf-cli/bin/cmds/helper/check-environment.ts
@@ -1,0 +1,76 @@
+import * as path from "path";
+import * as semver from "semver";
+import { Errors } from "../../../lib/errors";
+import { exec } from "../../../lib/util";
+
+function throwIfLowerVersion(
+  command: string,
+  language: string,
+  minVersion: string,
+  stdout: string
+) {
+  const version = semver.coerce(stdout);
+  if (!version || !semver.valid(version)) {
+    console.error(
+      Errors.Internal(
+        command,
+        `Unable to parse ${language} version "${stdout}"`
+      )
+    );
+    return;
+  }
+
+  if (semver.lt(version, minVersion)) {
+    console.error(
+      Errors.Usage(
+        command,
+        `${language} version "${version}" is not supported. Please upgrade to at least ${minVersion}`
+      )
+    );
+  }
+}
+
+function getBinaryVersion(
+  command: string,
+  binary: string,
+  versionCommand: string
+): Promise<string> {
+  try {
+    return exec(binary, [versionCommand], { env: process.env });
+  } catch (e) {
+    throw Errors.Usage(
+      command,
+      `Unable to run ${binary} ${versionCommand}, please check if ${binary} is installed: ${e}`
+    );
+  }
+}
+
+async function checkGoVersion(command: string) {
+  const out = await getBinaryVersion(command, "go", "version");
+  throwIfLowerVersion(command, "Go", "1.16.0", out);
+}
+
+async function checkNodeVersion(command: string) {
+  const out = await getBinaryVersion(command, "node", "--version");
+  throwIfLowerVersion(command, "Node.js", "12.16.0", out);
+}
+
+export async function checkEnvironment(
+  command: string,
+  projectPath = process.cwd()
+) {
+  await checkNodeVersion(command);
+
+  let language: string | undefined;
+  try {
+    const cdktfJson = require(path.resolve(projectPath, "cdktf.json"));
+    language = cdktfJson.language;
+  } catch {
+    // We can not detect the language
+  }
+
+  switch (language) {
+    case "go":
+      await checkGoVersion(command);
+  }
+}

--- a/packages/cdktf-cli/bin/cmds/init.ts
+++ b/packages/cdktf-cli/bin/cmds/init.ts
@@ -2,6 +2,7 @@ import yargs from "yargs";
 import { terraformCheck } from "./helper/terraform-check";
 import { displayVersionMessage } from "./helper/version-check";
 import { checkForEmptyDirectory, runInit, templates } from "./helper/init";
+import { checkEnvironment } from "./helper/check-environment";
 
 // eslint-disable-next-line @typescript-eslint/no-var-requires
 const pkg = require("../../package.json");
@@ -49,6 +50,7 @@ class Command implements yargs.CommandModule {
   public async handler(argv: any) {
     await terraformCheck();
     await displayVersionMessage();
+    await checkEnvironment("init");
 
     checkForEmptyDirectory(".");
 

--- a/packages/cdktf-cli/bin/cmds/list.ts
+++ b/packages/cdktf-cli/bin/cmds/list.ts
@@ -5,6 +5,7 @@ import { readConfigSync } from "../../lib/config";
 import { renderInk } from "./helper/render-ink";
 import { displayVersionMessage } from "./helper/version-check";
 import { throwIfNotProjectDirectory } from "./helper/check-directory";
+import { checkEnvironment } from "./helper/check-environment";
 
 const config = readConfigSync();
 
@@ -29,6 +30,7 @@ class Command implements yargs.CommandModule {
   public async handler(argv: any) {
     throwIfNotProjectDirectory("list");
     await displayVersionMessage();
+    await checkEnvironment("list");
     const command = argv.app;
     const outdir = argv.output;
 

--- a/packages/cdktf-cli/bin/cmds/synth.ts
+++ b/packages/cdktf-cli/bin/cmds/synth.ts
@@ -6,6 +6,7 @@ import { renderInk } from "./helper/render-ink";
 import * as fs from "fs-extra";
 import { displayVersionMessage } from "./helper/version-check";
 import { throwIfNotProjectDirectory } from "./helper/check-directory";
+import { checkEnvironment } from "./helper/check-environment";
 
 const config = readConfigSync();
 
@@ -41,6 +42,7 @@ class Command implements yargs.CommandModule {
   public async handler(argv: any) {
     throwIfNotProjectDirectory("synth");
     await displayVersionMessage();
+    await checkEnvironment("synth");
     const command = argv.app;
     const outdir = argv.output;
     const jsonOutput = argv.json;

--- a/packages/cdktf-cli/package.json
+++ b/packages/cdktf-cli/package.json
@@ -64,7 +64,7 @@
     "open": "^7.0.4",
     "parse-gitignore": "^1.0.1",
     "react": "<17.0.0",
-    "semver": "^7.3.2",
+    "semver": "^7.3.5",
     "sscaff": "^1.2.0",
     "stream-buffers": "^3.0.2",
     "strip-ansi": "^6.0.0",

